### PR TITLE
feat(versioning): drop new entries where newer versions already exist in the DB

### PR DESCRIPTION
**Pull request by louise-poole**
Important decisions made:
- The versioning logic assumes data is inserted as block-scoped data. Therefore if the db has an update for a specific entity id at a given block, it is assumed all updates for that entity have been inserted at that block
- Due to the above, db data inserts/updates much be batched by block for dynamic/versioned tables. For the most part this is already how things are done on the indexer with one exception: new contracts.
- `upsert_contract` used to both insert a new contract and update the dynamic tables related to it. This makes it difficult to batch all changes done to the dynamic tables (slots, balance, code etc) into one db interaction as these same tables are also updated by `update_contracts`. Therefore `upsert_contract` was changed to `insert_contract` and only inserts the static contract values. It is expected that `update_contract` be subsequently called with the dynamic values for them to be inserted for new contracts. 

NOTE - it is expected that for a well designed substream package we will only get a single message for a given contract: either a creation with the latest state values, or an update with the latest state values. However if for any reason a substream doesn't abide by this and has both a creation and update event for the same contract on the same block, our decisions made above will allow us to predictably handle the scenario.

### CODE REVIEW

File: tycho-storage/src/postgres/versioning.rs [hunk 1]
```diff
@@ -302,6 +377,18 @@ fn set_partitioned_versioning_attributes<N: PartitionedVersionedRow>(
...
+                if let Some(latest_row) = latest.get(id) {
+                    if latest_row.get_valid_from() > *delete_version {
+                        trace!(
+                            "Skipping delete for {:?} since it's older than the latest version",
+                            id
+                        );
+                        latest.remove(id);
+                        continue;
+                    }
+                }
```
Comment (kayibal) [inline tycho-storage/src/postgres/versioning.rs]: Did you think this case properly through? E.g. with attributes that oscillate between being deleted and being present...

**Review by kayibal — APPROVED** [review]
Thanks this looks good, just one q

File: tycho-storage/src/postgres/versioning.rs [hunk 2]
```diff
@@ -282,6 +355,19 @@ fn set_partitioned_versioning_attributes<N: PartitionedVersionedRow>(
...
                 let mut row = r.clone();
+
+                // Skip if new version is older than the latest version in the db
+                if let Some(latest_row) = latest.get(&id) {
+                    if latest_row.get_valid_from() > row.get_valid_from() {
+                        trace!(
+                            "Skipping update for {:?} since it's older than the latest version",
+                            id
+                        );
+                        latest.remove(&id);
```
Comment (zizou0x) [inline tycho-storage/src/postgres/versioning.rs]: I think this breaks if we have two old updates for the same ids. We would ignore the first one and remove it from `latest` but the next one wouldn't be ignored because `latest.get(&id)` would be `None`

**Review by zizou0x — COMMENTED** [review]
Looks good but I think there are some edge cases that aren't properly handled yet

I think we should add more extreme unit test cases to make sure we can handle them

File: tycho-storage/src/postgres/versioning.rs (hunk 1 above)
Comment (louise-poole) [reply tycho-storage/src/postgres/versioning.rs]: Fixed it [here](https://github.com/propeller-heads/tycho-indexer/pull/546/commits/735928d9b2737683aaa9b0d403c131367f37b3d1)

File: tycho-indexer/src/extractor/protocol_extractor.rs [hunk 3]
```diff
@@ -1410,6 +1414,7 @@ impl ExtractorGateway for ExtractorPgGateway {
         }
 
         // Insert component balance changes
+        dbg!(&component_balance_changes);
```
Comment (zizou0x) [inline tycho-indexer/src/extractor/protocol_extractor.rs]: Please remove this

File: tycho-storage/src/postgres/contract.rs [hunk 4]
```diff
@@ -995,12 +995,11 @@ impl PostgresGateway {
...
 
-    /// Upsert contract
+    /// Insert contract
     ///
-    /// Inserts a contract or updates it if it already exists. It will not update
-    /// contract balance or contract code if they already exist though. Since a separate
+    /// Inserts a contract. It will not insert contract code, slots or balance since a separate
     /// method exists for updating these related components.
-    pub async fn upsert_contract(
+    pub async fn insert_contract(
```
Comment (zizou0x) [inline tycho-storage/src/postgres/contract.rs:1001]: Isn't this still an upsert if we `do_update()`?

I would maybe rename to `upsert_account` instead?

File: tycho-storage/src/postgres/orm.rs [hunk 5]
```diff
@@ -1620,13 +1612,17 @@ impl PartitionedVersionedRow for NewSlot {
...
-                    .and(contract_storage::slot.eq_any(&slots))
-                    .and(contract_storage::valid_to.eq(MAX_TS)),
+                    .and(contract_storage::slot.eq_any(&slots)),
             )
+            .distinct_on((contract_storage::account_id, contract_storage::slot))
+            .order_by((
+                contract_storage::account_id,
+                contract_storage::slot,
+                contract_storage::valid_to.desc(),
+            ))
```
Comment (zizou0x) [inline tycho-storage/src/postgres/orm.rs:1639]: Isn't this completely equivalent to what was there before?

File: tycho-storage/src/postgres/versioning.rs [hunk 6]
```diff
@@ -218,6 +231,37 @@ where
...
+        .map(|row| row.get_entity_id())
+        .collect();
+    let latest_db_versions = S::latest_versions_by_ids(entity_ids.clone(), conn)
+        .await
+        .map_err(PostgresError::from)?
+        .iter()
+        .map(|row| (row.get_entity_id(), row.get_valid_from()))
+        .collect::<HashMap<N::EntityId, N::Version>>();
+
+    // Filter out entries that are older than the latest version in the db
```
Comment (zizou0x) [inline tycho-storage/src/postgres/versioning.rs]: Older or same ts

File: tycho-storage/src/postgres/versioning.rs [hunk 7]
```diff
@@ -266,31 +319,92 @@ type ArchivedRows<N> = Vec<N>;
...
+/// This function handles the versioning logic for partitioned tables by:
+/// 1. Setting up archive versions for entities that are being updated
+/// 2. Marking rows as deleted when deletion entries are encountered
+/// 3. Maintaining the latest version of each entity
+/// 4. Skipping updates or deletions that are older than versions already in the database
+///
+/// The function maintains a collection of latest rows, archived rows, and deleted entity IDs.
+/// For each entity:
+/// - If a newer version arrives, the current version is archived and the new one becomes the latest
+/// - If a deletion arrives, the current version is archived, marked as deleted, and added to the
```
Comment (zizou0x) [inline tycho-storage/src/postgres/versioning.rs]:
```suggestion
/// - If a new deletion arrives, the current version is archived, marked as deleted, and added to the
```

File: tycho-storage/src/postgres/versioning.rs [hunk 8]
```diff
@@ -266,31 +319,92 @@ type ArchivedRows<N> = Vec<N>;
...
+///   deleted IDs
+/// - If an update or deletion is older than the current version in the database, it's skipped
```
Comment (zizou0x) [inline tycho-storage/src/postgres/versioning.rs]:
```suggestion
/// - If an update or deletion is older or equal than the current version in the database, it's skipped
```

File: tycho-storage/src/postgres/versioning.rs [hunk 9]
```diff
@@ -266,31 +319,92 @@ type ArchivedRows<N> = Vec<N>;
...
+                }
+
                 if let Some(mut prev) = latest.remove(&id) {
                     prev.archive(&mut row);
                     archived.push(prev);
                 }
                 // If it's updated after being deleted, it doesn't need to be marked as deleted
                 deleted.remove(&id);
+                // If it is updated after being skipped, it doesn't need to be skipped
+                to_skip.remove(&id);
```
Comment (zizou0x) [inline tycho-storage/src/postgres/versioning.rs]: Can this happen? I like it like this, it's easy and robust, but I just wonder

**Review by zizou0x — APPROVED** [review]
Looks good to me, just some small comments. It would be nice to get one more approval just to be sure

File: tycho-storage/src/postgres/orm.rs (hunk 5 above)
Comment (zizou0x) [reply tycho-storage/src/postgres/orm.rs:1639]: I think this is to handle deleted rows right?

File: tycho-storage/src/postgres/orm.rs [hunk 10]
```diff
@@ -455,13 +455,17 @@ impl PartitionedVersionedRow for NewComponentBalance {
...
-                    .and(component_balance::token_id.eq_any(&token_ids))
-                    .and(component_balance::valid_to.eq(MAX_TS)),
+                    .and(component_balance::token_id.eq_any(&token_ids)),
             )
+            .distinct_on((component_balance::protocol_component_id, component_balance::token_id))
+            .order_by((
+                component_balance::protocol_component_id,
+                component_balance::token_id,
+                component_balance::valid_to.desc(),
+            ))
```
Comment (kayibal) [inline tycho-storage/src/postgres/orm.rs:473]: Very worried about the performance here... I guess we had an index on valid_to and now this needs to query all partitions, aggregate the data in a temp table, sort it and then remove dupes...

After reading the all changes I understand you want to also get the latest deleted row but I still don't understand why.

File: tycho-storage/src/postgres/contract.rs [hunk 11]
```diff
@@ -1034,68 +1033,27 @@ impl PostgresGateway {
...
-                .map_err(|err| {
-                    error!("Failed inserting account");
-                    storage_error_from_diesel(err, "Account", &hex_addr, None)
-                })?
-        };
+        use schema::account::dsl;
+        diesel::insert_into(schema::account::table)
+            .values(new_contract.new_account())
+            .on_conflict(on_constraint("account_chain_id_address_key"))
+            .do_update()
```
Comment (kayibal) [inline tycho-storage/src/postgres/contract.rs]: Since it is called insert, I think this should error if the contract already exists.

File: tycho-storage/src/postgres/contract.rs (hunk 4 above)
Comment (louise-poole) [reply tycho-storage/src/postgres/contract.rs:1001]: Ah true! Will change this back. Technically the stuff it updates is static (title, creation_tx etc). I'm not sure why we consider this updatable. But won't get into that in this PR - I'll just revert the rename.
Edit: a later comment by Alan suggests removing the update part so I'll do that instead.

File: tycho-storage/src/postgres/orm.rs (hunk 5 above)
Comment (louise-poole) [reply tycho-storage/src/postgres/orm.rs:1639]: Yes - this is to handle deleted entities. Previously this would fetch only currently valid versions, but for our versioning logic to work we need to fetch the latest versions (irrespective of if they're considered current or not).

File: tycho-storage/src/postgres/orm.rs (hunk 10 above)
Comment (louise-poole) [reply tycho-storage/src/postgres/orm.rs:473]: I assume after the slack thread it's clear?
I've added a PERF comment here so that it's clear where we need to return to for the performance improvements.
